tools/cndump/kvset_dump.c

```diff
@@ -0,0 +1,385 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/*
+ * Copyright (C) 2022 Micron Technology, Inc.  All rights reserved.
+ */
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include <hse/hse.h>
+
+#include <hse_ikvdb/diag_kvdb.h>
+#include <hse_ikvdb/ikvdb.h>
+#include <hse_ikvdb/omf_kmd.h>
+
+#include <hse_util/parse_num.h>
+#include <hse_util/fmt.h>
+
+#include <mpool/mpool_structs.h>
+#include <mpool/mpool.h>
+
+#include <cn/omf.h>
+
+#include "cndb_reader.h"
+#include "cndb_record.h"
+#include "fatal.h"
+#include "globals.h"
+#include "commands.h"
+
+/* command line options for kvset sub-command */
+struct opts {
+    uint64_t kvset_id;          // CNDB ID of kvset to dump
+    const char *home;           // KVDB home dir
+};
+
+static struct opts opts;
+
+struct dump_mblock {
+    struct mblock_props props;
+    void               *data;
+};
+
+struct dump_kvset {
+    struct cndb_rec     rec;
+    struct dump_mblock  hblk;
+    struct dump_mblock *kblkv;
+    struct dump_mblock *vblkv;
+};
+
+static void
+help(void)
+{
+    printf("usage: %s %s [options] <kvdb_home> <kvset_id>\n", progname, KVSET_COMMAND_NAME);
+    printf(
+        "  -h            print help\n"
+        "  -v            verbose\n"
+        "  <kvdb_home>   KVDB home directory\n"
+        "  <kvset_id>    kvset ID (from the CNDB log)\n");
+
+    if (!global_opts.verbose) {
+        printf("use '-hv' for more detail\n");
+        return;
+    }
+
+    printf("\nDump a KVSET on standard output. Kvset IDs can be seen in the\n"
+        "output of '%s %s'.\n", progname, CNDB_COMMAND_NAME);
+}
+
+static void
+parse_args(int argc, char **argv)
+{
+    const char *arg;
+    int c;
+
+    while ((c = getopt(argc, argv, "+:hv")) != -1) {
+        switch (c) {
+        case 'h':
+            global_opts.help = true;
+            break;
+        case 'v':
+            global_opts.verbose = true;
+            break;
+        case ':':
+            syntax("option -%c requires a parameter", optopt);
+            break;
+        default:
+            syntax("invalid %s option: -%c", KVSET_COMMAND_NAME, optopt);
+            break;
+        }
+    }
+
+    if (global_opts.help) {
+        help();
+        exit(0);
+    }
+
+    if (optind == argc)
+        syntax("missing <kvdb_home> parameter");
+
+    opts.home = argv[optind++];
+
+    if (optind == argc)
+        syntax("missing <kvset_id> parameter");
+
+    arg = argv[optind++];
+    if (parse_u64(arg, &opts.kvset_id))
+        syntax("invalid kvset ID: '%s' (cannot convert to uint64_t)", arg);
+
+    if (optind != argc)
+        syntax("unexpected parameter: '%s'", argv[optind]);
+}
+
+static char *
```

Comment:
you can mark this `HSE_RETURNS_NONNULL`

Reply:
If caller passes in addr of NULL ptr and len==0, it will return NULL.